### ISSUE TITLE
`/api/v4/contests/2/submissions/1/source-code` May not work properly when dealing with Chinese

### ISSUE THREAD

**Issue by Dup4** [DOMjudge/domjudge#1394]
NOTE: If this is a _discussion starter_, you need any _installation help_ or have a _question_ on how to accomplish something,
rather post at [our discussion channel](https://www.domjudge.org/chat) or send an email to our
[DOMjudge-devel mailinglist](https://www.domjudge.org/mailman/listinfo/domjudge-devel)
instead of filing an issue here.

### Description of the problem
> Replace this line with a short description.

Our users include Chinese users, so there may be Chinese comments in the code.

`/api/v4/contests/2/submissions/1/source-code` will return data with the following structure:

```json
[
  {
    "id": "string",
    "submission_id": "string",
    "filename": "string",
    "source": "string"
  }
]
```

And the `source` field seems to be base64 encoded.

So we need to base64 decode the field when calling this api to get the source code.

But when there are Chinese comments in the code, the decoded source code does not look as expected.

For example, the source code is:

```cpp
/*********************************************************************
    程序名:
    日期: 2021-12-12 13:33
    说明:
*********************************************************************/
#include <stdio.h>

int main() {
	int n, i;
	scanf("%d", &n);
	for (i = 0; i < n; i++) {
		printf("Welcome to HZNU\n");
	}
	return 0;
}
```

The value of the `source` field obtained from the interface is:

```base64
LyoqKioqKioqKioqKioqKioqKioqKioqKioqKioqKioqKioqKioqKioqKioqKioqKioqKioqKioqKioqKioqKioqKioqKg0KICAgILPM0PLD+zoNCiAgICDI1cbaOiAyMDIxLTEyLTEyIDEzOjMzDQogICAgy7XD9zoNCioqKioqKioqKioqKioqKioqKioqKioqKioqKioqKioqKioqKioqKioqKioqKioqKioqKioqKioqKioqKioqKioqKioqKi8NCiNpbmNsdWRlIDxzdGRpby5oPg0KDQppbnQgbWFpbigpIHsNCglpbnQgbiwgaTsNCglzY2FuZigiJWQiLCAmbik7DQoJZm9yIChpID0gMDsgaSA8IG47IGkrKykgew0KCQlwcmludGYoIldlbGNvbWUgdG8gSFpOVVxuIik7DQoJfQ0KCXJldHVybiAwOw0KfQ0K
```

Then we decode it:

![image](https://user-images.githubusercontent.com/38343778/146925875-705613a2-461a-454c-a93b-1f80d9372542.png)

We can see that the Chinese part is garbled.

### Your environment
> Include details about your installation here.
> - DOMjudge version (e.g. 7.0.0 or a github commit hash)
> - Operating system / Linux distribution and version (e.g. Ubuntu 18.04)
> - Webserver (e.g. Apache or nginx)

- 8.0.0DEV
- Ubuntu 20.04
- Nginx

BTW, I use docker to run domserver

### Steps to reproduce
> Replace this with a description how we can reproduce your bug.
> - Step 1
> - Step 2
> - Step 3

### Expected behaviour
> Replace this line with what you would expect to happen.

### Actual behaviour
> Replace this line with what happens instead.

### Any other information that you want to share?
> Please include webserver, symfony and judgedaemon log snippets here as appropriate.
> Screenshots may help in case of UI bugs.

**Comment by nickygerritsen** [comment]
Hmmm we just use `base64_decode` (https://github.com/DOMjudge/domjudge/blob/main/webapp/src/Controller/API/SubmissionController.php#L537) so I wonder what goes wrong. Maybe the data is not returned or stored in MySQL correctly? Does it look OK in the jury interface if you view the source code? If you request `/api/v4/contests/2/submissions/1/files` do you get a ZIP with the file? If you unpack that ZIP are the contests correct?

**Comment by Dup4** [comment]
I use `/api/v4/contests/2/submissions/1/files` to get the code file.

- If it is opened with `utf-8` encoding, it is still garbled.
- If it is opened with `gb2312` encoding, then it is normal.

So I realized that it might be an encoding problem.

So I decoded the source field with `base64` and then decoded it with `gb2312` to get the normal code.

By the way, why use `gb2312` encoding instead of `utf-8` encoding.
